cpp/ql/lib/semmle/code/cpp/rangeanalysis/new/internal/semantic/analysis/RangeAnalysisConstantSpecific.qll

```diff
@@ -7,6 +7,57 @@ private import RangeAnalysisStage
 private import semmle.code.cpp.rangeanalysis.new.internal.semantic.analysis.FloatDelta
 
 module CppLangImplConstant implements LangSig<FloatDelta> {
+  private newtype TSemReason =
+    TSemNoReason() or
+    TSemCondReason(SemGuard guard) or
+    TSemTypeReason()
+
+  /**
+   * A reason for an inferred bound. This can either be `CondReason` if the bound
+   * is due to a specific condition, or `NoReason` if the bound is inferred
+   * without going through a bounding condition.
+   */
+  abstract class SemReason extends TSemReason {
+    /** Gets a textual representation of this reason. */
+    abstract string toString();
+
+    bindingset[this, reason]
+    abstract SemReason combineWith(SemReason reason);
+  }
+
+  /**
+   * A reason for an inferred bound that indicates that the bound is inferred
+   * without going through a bounding condition.
+   */
+  class SemNoReason extends SemReason, TSemNoReason {
+    override string toString() { result = "NoReason" }
+
+    override SemReason combineWith(SemReason reason) { result = reason }
+  }
+
+  /** A reason for an inferred bound pointing to a condition. */
+  class SemCondReason extends SemReason, TSemCondReason {
+    /** Gets the condition that is the reason for the bound. */
+    SemGuard getCond() { this = TSemCondReason(result) }
+
+    override string toString() { result = this.getCond().toString() }
+
+    bindingset[this, reason]
+    override SemReason combineWith(SemReason reason) {
+      if reason instanceof SemTypeReason then result instanceof SemTypeReason else result = this
```

Comment:
Why is this the correct thing to do?

Reply:
If the reason is a type reason we need to propagate this upwards as the final `SemReason` we report needs to be a `TypeReason` if the bound is influenced by type bounds (so that a query can ignore them if it wants to). However, if this is _not_ a type reason, then we just what to do the same thing as the library did before we introduced `TypeReason`s. So whenever the old library did:
```ql
... and reason = r1
or
... and reason = r2
```
and inferred a bound based on two recursive calls that produced reasons `r1` and `r2` respectively), we now "taint" the reason and mark it as a `TypeBound` if one of the reasons is a type-reason like so:
```ql
... and reason = r1.combineWith(r2)
or
... and reason = r2.combineWith(r1)
```

Does that make sense?

Reply:
Why can't we just propagate all the reasons, with the `TypeBound`s being among them?

Reply:
So you mean instead of:
```ql
... and reason = r1.combineWith(r2)
or
... and reason = r2.combineWith(r1)
```
we do something like:
```ql
... and reason = [r1, r2]
or
... and reason = [r1, r2]
```
?

Reply:
I think I'm missing some subtlety here. It seems that we were not propagating all the bounds anyway? At least that's my impression when I read something like: `(if r1 instanceof SemNoReason then reason = r2 else reason = r1)`.

Reply:
> if we ever got in a situation where we had to pick between two CondReasons we just picked an arbitrary one.

So for `(if r1 instanceof SemNoReason then reason = r2 else reason = r1)`, arbitrary means the first one that is not `SemNoReason`. Correct?

Reply:
Correct

Reply:
And now we pick the same "arbitrary" reason as before, but if any of them is a `TypeReason` then we definitely report a `TypeReason` back.

Reply:
Got it. I think it would be worth adding a comment to `if reason instanceof SemTypeReason then result instanceof SemTypeReason else result = this` that explains this.

Reply:
Good idea. Will do!

Reply:
Fixed in 86dbb507fd3579858688d37888157d31f242161e.